fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+    let mut elements = vec![];
+    let mut last_non_blank = None;
+    let mut common_indent = None;
+
+    ps.skip_blank_inline();
+
+    let mut text_element_role = if ps.skip_eol() {
+        ps.skip_blank_block();
+        TextElementPosition::LineStart
+    } else {
+        TextElementPosition::InitialLineStart
+    };
+
+    while ps.ptr < ps.length {
+        if ps.is_current_byte(b'{') {
+            if text_element_role == TextElementPosition::LineStart {
+                common_indent = Some(0);
+            }
+            let exp = get_placeable(ps)?;
+            last_non_blank = Some(elements.len());
+            elements.push(PatternElementPointers::Placeable(exp));
+            text_element_role = TextElementPosition::Continuation;
+        } else {
+            let slice_start = ps.ptr;
+            let mut indent = 0;
+            if text_element_role == TextElementPosition::LineStart {
+                indent = ps.skip_blank_inline();
+                if ps.ptr >= ps.length {
+                    break;
+                }
+                let b = ps.source[ps.ptr];
+                if indent == 0 {
+                    if b != b'\n' {
+                        break;
+                    }
+                } else if !ps.is_byte_pattern_continuation(b) {
+                    ps.ptr = slice_start;
+                    break;
+                }
+            }
+            let (start, end, text_element_type, termination_reason) = get_text_slice(ps)?;
+            if start != end {
+                if text_element_role == TextElementPosition::LineStart
+                    && text_element_type == TextElementType::NonBlank
+                {
+                    if let Some(common) = common_indent {
+                        if indent < common {
+                            common_indent = Some(indent);
+                        }
+                    } else {
+                        common_indent = Some(indent);
+                    }
+                }
+                if text_element_role != TextElementPosition::LineStart
+                    || text_element_type == TextElementType::NonBlank
+                    || termination_reason == TextElementTermination::LineFeed
+                {
+                    if text_element_type == TextElementType::NonBlank {
+                        last_non_blank = Some(elements.len());
+                    }
+                    elements.push(PatternElementPointers::TextElement(
+                        slice_start,
+                        end,
+                        indent,
+                        text_element_role,
+                    ));
+                }
+            }
+
+            text_element_role = match termination_reason {
+                TextElementTermination::LineFeed => TextElementPosition::LineStart,
+                TextElementTermination::CarriageReturn => TextElementPosition::Continuation,
```

Comment:
`CarriageReturn` should really be called `CRLF`, as we don't support the lone `CR` as a valid new line. Please add docs which explain why `TextElementTermination::CarriageReturn` results in something else than `TextElementTermination::LineFeed` does. In particular, AFAIUI, the quirk of the zero-copy design is that `a\r\n    b` becomes three `TextElements`: `a`, `\n` and `b`.

Reply:
(I understand that when serialized to JSON, these three `TextElements` become a single one, which is very neat.)